papaya-core/src/main/java/pl/edu/agh/papaya/service/jira/JiraOAuthClient.java

```diff
@@ -0,0 +1,129 @@
+            return Optional.of(authorizationURL.toString());
+        } catch (MalformedURLException e) {
+            log.error("authorizationUrl is incorrect", e);
+            return Optional.empty();
+        }
+    }
+
+    Optional<String> getAccessToken(String jiraVerCode, String tmpToken, String privateKey) {
+        try {
+            JiraOAuthGetAccessToken oAuthAccessToken =
+                    oAuthTokenFactory.getJiraOAuthGetAccessToken(jiraVerCode, tmpToken, privateKey);
+            OAuthCredentialsResponse response = oAuthAccessToken.execute();
+            return Optional.of(response.token);
+        } catch (NoSuchAlgorithmException | InvalidKeySpecException | IOException e) {
+            log.error("getAccessToken failed", e);
+            return Optional.empty();
+        }
+    }
+
+    Optional<JSONArray> makeAuthorizedGetRequest(String jiraVerificationCode,
+            String accessToken, String endPoint, String privateKey) {
+        try {
+            OAuthParameters parameters = getParameters(jiraVerificationCode, accessToken, privateKey);
+            HttpResponse response =
+                    getResponseFromUrl(new GenericUrl(new URL(jiraHome, endPoint).toString()), parameters);
+            return Optional.of(parseResponse(response));
+        } catch (NoSuchAlgorithmException | IOException | InvalidKeySpecException e) {
+            log.error("makeAuthorizedGetRequest failed", e);
+            return Optional.empty();
+        }
+    }
+
+    private OAuthParameters getParameters(String jiraVerCode, String tmpToken, String privateKey)
+            throws NoSuchAlgorithmException, InvalidKeySpecException {
+        JiraOAuthGetAccessToken oAuthAccessToken =
+                oAuthTokenFactory.getJiraOAuthGetAccessToken(jiraVerCode, tmpToken, privateKey);
+        oAuthAccessToken.verifier = jiraVerCode;
+        return oAuthAccessToken.createParameters();
+    }
+
+    private HttpResponse getResponseFromUrl(GenericUrl jiraUrl, OAuthParameters parameters) throws IOException {
+        HttpRequestFactory requestFactory = new NetHttpTransport().createRequestFactory(parameters);
+        HttpRequest request = requestFactory.buildGetRequest(jiraUrl);
+        return request.execute();
+    }
+
+    private JSONArray parseResponse(HttpResponse response) throws IOException {
+        Scanner scanner = new Scanner(response.getContent(), StandardCharsets.UTF_8);
+        StringBuilder result = new StringBuilder();
+        while (scanner.hasNext()) {
+            result.append(scanner.next());
+        }
+
+        if (result.charAt(0) == '{') {
+            result.insert(0, "[");
+            result.append("]");
+        }
+
+        return new JSONArray(result.toString());
+    }
+
+    static class JiraOAuthClientFactory {
+
+        static Optional<JiraOAuthClient> getJiraOAuthClient(URL jiraHome) {
+            try {
+                URL jiraHomeWithTrailingSlash = new URL(String.format("%s/", jiraHome.toString()));
```

Comment:
1. co jeśli `jiraHome` już kończy się slashem?
2. czemu używamy `String.format`?
3. czemu to jest w nested klasie, nie może być jako static factory method?

Reply:
@MrMarosis  @kjarosh Tu nie nastąpił _konsensus_ jeszcze?  Na jutro nastąpi? (Mnie czeka kolejny rebase z piekła na zmiany stąd)

Reply:
1. Dodano obsługę dla takiego przypadku. To była jakaś naleciałość z założenia, że nie ma `jiraHome` typu `.../jira`
2. Zamieniono wywołanie `String.format` na `StringBuilder`, co pozwoli na zmniejszenie czasu operacji o rząd wielkości. Nie jestem pewien o co chodzi, chyba uwaga ta dotyczy kwestii wydajności
3. Zamieniono na metodę. Podążałem za przykładami z jakiegoś forum. Spodobała mi się idea dedykowanej klasy